### ISSUE TITLE
`kapp` deletes `ServiceAccount` before `PackageInstall`

### ISSUE THREAD

**Issue by mamachanko** [k14s/kapp#579]
**What steps did you take:**

Deploy an `ns`, `pkg`, `pkgi`, `SA` and RBAC with kapp:
```shell
❯ ytt -f debug.yaml | kapp d -a debug -f- -y -c
Target cluster *****

@@ create namespace/debug (v1) cluster @@
...
@@ create package/my.example.com.1.0.0 (data.packaging.carvel.dev/v1alpha1) namespace: debug @@
...
@@ create clusterrole/debug-cluster-admin-role (rbac.authorization.k8s.io/v1) cluster @@
...
@@ create clusterrolebinding/debug-cluster-admin-role-binding (rbac.authorization.k8s.io/v1) cluster @@
...
@@ create serviceaccount/debug-sa (v1) namespace: debug @@
...
@@ create packageinstall/example-install (packaging.carvel.dev/v1alpha1) namespace: debug @@
...
@@ create secret/example-install-values (v1) namespace: debug @@
...

Changes

Namespace  Name                              Kind                Age  Op      Op st.  Wait to    Rs  Ri
(cluster)  debug                             Namespace           -    create  -       reconcile  -   -
^          debug-cluster-admin-role          ClusterRole         -    create  -       reconcile  -   -
^          debug-cluster-admin-role-binding  ClusterRoleBinding  -    create  -       reconcile  -   -
debug      debug-sa                          ServiceAccount      -    create  -       reconcile  -   -
^          example-install                   PackageInstall      -    create  -       reconcile  -   -
^          example-install-values            Secret              -    create  -       reconcile  -   -
^          my.example.com.1.0.0              Package             -    create  -       reconcile  -   -

Op:      7 create, 0 delete, 0 update, 0 noop, 0 exists
Wait to: 7 reconcile, 0 delete, 0 noop

3:13:42PM: ---- applying 2 changes [0/7 done] ----
3:13:42PM: create clusterrole/debug-cluster-admin-role (rbac.authorization.k8s.io/v1) cluster
3:13:42PM: create namespace/debug (v1) cluster
3:13:42PM: ---- waiting on 2 changes [0/7 done] ----
3:13:42PM: ok: reconcile clusterrole/debug-cluster-admin-role (rbac.authorization.k8s.io/v1) cluster
3:13:42PM: ok: reconcile namespace/debug (v1) cluster
3:13:42PM: ---- applying 3 changes [2/7 done] ----
3:13:42PM: create secret/example-install-values (v1) namespace: debug
3:13:42PM: create clusterrolebinding/debug-cluster-admin-role-binding (rbac.authorization.k8s.io/v1) cluster
3:13:43PM: create serviceaccount/debug-sa (v1) namespace: debug
3:13:43PM: ---- waiting on 3 changes [2/7 done] ----
3:13:44PM: ok: reconcile serviceaccount/debug-sa (v1) namespace: debug
3:13:44PM: ok: reconcile secret/example-install-values (v1) namespace: debug
3:13:44PM: ok: reconcile clusterrolebinding/debug-cluster-admin-role-binding (rbac.authorization.k8s.io/v1) cluster
3:13:44PM: ---- applying 2 changes [5/7 done] ----
3:13:44PM: create package/my.example.com.1.0.0 (data.packaging.carvel.dev/v1alpha1) namespace: debug
3:13:45PM: create packageinstall/example-install (packaging.carvel.dev/v1alpha1) namespace: debug
3:13:45PM: ---- waiting on 2 changes [5/7 done] ----
3:13:45PM: ok: reconcile packageinstall/example-install (packaging.carvel.dev/v1alpha1) namespace: debug
3:13:45PM: ok: reconcile package/my.example.com.1.0.0 (data.packaging.carvel.dev/v1alpha1) namespace: debug
3:13:45PM: ---- applying complete [7/7 done] ----
3:13:45PM: ---- waiting complete [7/7 done] ----

Succeeded
```

Uninstall the app:
```shell
❯ kapp delete -a debug -y
Target cluster 'https://applaud-revenue-56f55d6f.hcp.eastus.azmk8s.io:443' (nodes: aks-agentpool-17938520-vmss000001, 2+)

Changes

Namespace  Name                              Kind                Age  Op      Op st.  Wait to  Rs  Ri
(cluster)  debug                             Namespace           54s  delete  -       delete   ok  -
^          debug-cluster-admin-role          ClusterRole         54s  delete  -       delete   ok  -
^          debug-cluster-admin-role-binding  ClusterRoleBinding  54s  delete  -       delete   ok  -
debug      debug-sa                          ServiceAccount      54s  delete  -       delete   ok  -
^          example-install                   PackageInstall      52s  delete  -       delete   ok  -
^          example-install-values            Secret              54s  delete  -       delete   ok  -
^          my.example.com.1.0.0              Package             52s  delete  -       delete   ok  -

Op:      0 create, 7 delete, 0 update, 0 noop, 0 exists
Wait to: 0 reconcile, 7 delete, 0 noop

3:14:36PM: ---- applying 4 changes [0/7 done] ----
3:14:36PM: delete secret/example-install-values (v1) namespace: debug
3:14:36PM: delete packageinstall/example-install (packaging.carvel.dev/v1alpha1) namespace: debug
3:14:36PM: delete namespace/debug (v1) cluster
3:14:36PM: delete package/my.example.com.1.0.0 (data.packaging.carvel.dev/v1alpha1) namespace: debug
3:14:36PM: ---- waiting on 4 changes [0/7 done] ----
3:14:36PM: ongoing: delete packageinstall/example-install (packaging.carvel.dev/v1alpha1) namespace: debug
3:14:36PM:  ^ Waiting on finalizers: finalizers.packageinstall.packaging.carvel.dev/delete
3:14:36PM: ok: delete secret/example-install-values (v1) namespace: debug
3:14:36PM: ongoing: delete namespace/debug (v1) cluster
3:14:36PM:  ^ Deleting
3:14:36PM: ok: delete package/my.example.com.1.0.0 (data.packaging.carvel.dev/v1alpha1) namespace: debug
3:14:36PM: ---- waiting on 2 changes [2/7 done] ----
```

```yaml
# debug.yaml

#@ load("@ytt:data", "data")

#@ namespace = getattr(data.values, "namespace", "debug")

---
apiVersion: v1
kind: Namespace
metadata:
  name: #@ namespace

---
apiVersion: data.packaging.carvel.dev/v1alpha1
kind: Package
metadata:
  name: my.example.com.1.0.0
  namespace: #@ namespace
spec:
  version: 1.0.0
  refName: my.example.com
  template:
    spec:
      fetch:
        - inline:
            paths:
              resources.yaml: |
                #@ load("@ytt:data", "data")
                ---
                apiVersion: v1
                kind: Namespace
                metadata:
                  name: #@ data.values.namespace
                ---
                apiVersion: v1
                kind: ConfigMap
                metadata:
                  name: config
                  namespace: #@ data.values.namespace
                data:
                  message: it's fun
      template:
        - ytt:
            paths: [.]
      deploy:
        - kapp: { }

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: debug-cluster-admin-role
rules:
  - apiGroups:
      - '*'
    resources:
      - '*'
    verbs:
      - '*'

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: debug-cluster-admin-role-binding
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: debug-cluster-admin-role
subjects:
  - kind: ServiceAccount
    name: debug-sa
    namespace: #@ namespace

---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: debug-sa
  namespace: #@ namespace

---
apiVersion: packaging.carvel.dev/v1alpha1
kind: PackageInstall
metadata:
  name: example-install
  namespace: #@ namespace
spec:
  serviceAccountName: debug-sa
  packageRef:
    refName: my.example.com
    versionSelection:
      constraints: 1.0.0
  values:
    - secretRef:
        name: example-install-values

---
apiVersion: v1
kind: Secret
metadata:
  name: example-install-values
  namespace: #@ namespace
stringData:
  values.yaml: |
    namespace: example-test
```

**What happened:**
The deleteion of the kapp app is stuck on the `PackageInstall`, because its `SA` is already deleted:
```shell
❯ k get pkgi -n debug example-install -oyaml | yq .status
conditions:
  - message: Error (see .status.usefulErrorMessage for details)
    status: "True"
    type: DeleteFailed
friendlyDescription: 'Delete failed: Error (see .status.usefulErrorMessage for details)'
lastAttemptedVersion: 1.0.0
observedGeneration: 2
usefulErrorMessage: 'Preparing kapp: Getting service account: serviceaccounts "debug-sa" not found'
version: 1.0.0
```

**What did you expect:**

The kapp app smoothly deletes and cleans up.

**Anything else you would like to add:**

**Environment:**

- kapp version (use `kapp --version`): `0.52.0`
- kapp-controller version: `v0.38.4`
- OS (e.g. from `/etc/os-release`): macOS
- Kubernetes version (use `kubectl version`)
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:33:37Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.11", GitCommit:"6b66c9a34569cd5a6e330eda21d1d8bd792e164e", GitTreeState:"clean", BuildDate:"2022-06-21T17:14:21Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by praveenrewar** [comment]
Hi @mamachanko !
Can you try using the latest version of kapp-controller (v0.39.0 I think) as the version you have mentioned is using an old version of kapp which doesn't have the default ordering for these resources.

**Comment by mamachanko** [comment]
> Hi @mamachanko ! Can you try using the latest version of kapp-controller (v0.39.0 I think) as the version you have mentioned is using an old version of kapp which doesn't have the default ordering for these resources.

I can, but we are locked into `kapp-controller v0.38.4` via [Tanzu Cluster Essentials](https://network.tanzu.vmware.com/products/tanzu-cluster-essentials). Furthermore, I want to manage these resources with `kapp` from outside the cluster.

**Comment by kumaritanushree** [comment]
@mamachanko thanks for bringing up this issue. As we had discussion over slack, deleting namespace  before (Just the logical order created by kapp) serviceAccount and packageInstall(which is part of that namespace itself) seems causing issue. It seems kapp is not following the order to delete namespaced resources properly.  Will dig into it more and will try to resolve it ASAP.

**Comment by mamachanko** [comment]
As @kumaritanushree suggested, I can confirm that a custom set of change groups and rules successfully works around the issue:

```diff
diff --git a/debug.yaml b/debug.yaml
index 5a23fc01..9365ba0e 100644
--- a/debug.yaml
+++ b/debug.yaml
@@ -7,6 +7,8 @@ apiVersion: v1
 kind: Namespace
 metadata:
   name: #@ namespace
+  annotations:
+    kapp.k14s.io/change-group: ns
 
 ---
 apiVersion: data.packaging.carvel.dev/v1alpha1
@@ -76,6 +78,9 @@ kind: ServiceAccount
 metadata:
   name: debug-sa
   namespace: #@ namespace
+  annotations:
+    kapp.k14s.io/change-group: sa
+    kapp.k14s.io/change-rule: "delete before deleting ns"
 
 ---
 apiVersion: packaging.carvel.dev/v1alpha1
@@ -83,6 +88,8 @@ kind: PackageInstall
 metadata:
   name: example-install
   namespace: #@ namespace
+  annotations:
+    kapp.k14s.io/change-rule: "delete before deleting sa"
 spec:
   serviceAccountName: debug-sa
   packageRef:
@@ -99,6 +106,8 @@ kind: Secret
 metadata:
   name: example-install-values
   namespace: #@ namespace
+  annotations:
+    kapp.k14s.io/change-rule: "delete before deleting sa"
 stringData:
   values.yaml: |
     namespace: example-test
```

**Comment by praveenrewar** [comment]
Ah, sorry, missed the namespace part. 
@kumaritanushree I think we are missing a default change rule to delete resources belonging to a namespace before deleting the namespace itself.

**Comment by renuy** [comment]
Released as part of v0.53.0
